src/Orleans.Runtime/Placement/ResourceOptimizedPlacementDirector.cs

```diff
@@ -0,0 +1,257 @@
+            {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /// <summary>
+    /// Always returns a value [0-1]
+    /// </summary>
+    /// <returns>
+    /// score = cpu_weight * (cpu_usage / 100) +
+    ///         mem_usage_weight * (mem_usage / physical_mem) +
+    ///         mem_avail_weight * [1 - (mem_avail / physical_mem)]
+    ///         physical_mem_weight * (1 / (1024 * 1024 * physical_mem)
+    /// </returns>
+    /// <remarks>physical_mem is represented in [MB] to keep the result within [0-1] in cases of silos having physical_mem less than [1GB]</remarks>
+    private float CalculateScore(ResourceStatistics stats)
+    {
+        float normalizedCpuUsage = stats.CpuUsage.HasValue ? stats.CpuUsage.Value / 100f : 0f;
+
+        if (stats.TotalPhysicalMemory is { } physicalMemory && physicalMemory > 0)
+        {
+            float normalizedMemoryUsage = stats.MemoryUsage.HasValue ? stats.MemoryUsage.Value / physicalMemory : 0f;
+            float normalizedAvailableMemory = 1 - (stats.AvailableMemory.HasValue ? stats.AvailableMemory.Value / physicalMemory : 0f);
+            float normalizedPhysicalMemory = PhysicalMemoryScalingFactor * physicalMemory;
+
+            return _options.CpuUsageWeight * normalizedCpuUsage +
+                   _options.MemoryUsageWeight * normalizedMemoryUsage +
+                   _options.AvailableMemoryWeight * normalizedAvailableMemory +
+                   _options.AvailableMemoryWeight * normalizedPhysicalMemory;
+        }
+
+        return _options.CpuUsageWeight * normalizedCpuUsage;
+    }
+
+    public void RemoveSilo(SiloAddress address)
+         => siloStatistics.TryRemove(address, out _);
+
+    public void SiloStatisticsChangeNotification(SiloAddress address, SiloRuntimeStatistics statistics)
+        => siloStatistics.AddOrUpdate(
+            key: address,
+            addValue: new ResourceStatistics(
+                statistics.CpuUsage,
+                statistics.AvailableMemory,
+                statistics.MemoryUsage,
+                statistics.TotalPhysicalMemory,
+                statistics.IsOverloaded),
+            updateValueFactory: (_, _) =>
+            {
+                float estimatedCpuUsage = _cpuUsageFilter.Filter(statistics.CpuUsage);
+                float estimatedAvailableMemory = _availableMemoryFilter.Filter(statistics.AvailableMemory);
+                long estimatedMemoryUsage = _memoryUsageFilter.Filter(statistics.MemoryUsage);
+
+                return new ResourceStatistics(
+                    estimatedCpuUsage,
+                    estimatedAvailableMemory,
+                    estimatedMemoryUsage,
+                    statistics.TotalPhysicalMemory,
+                    statistics.IsOverloaded);
+            });
+
+    // details: https://www.ledjonbehluli.com/posts/orleans_resource_placement_kalman/
+    private sealed class DualModeKalmanFilter<T> where T : unmanaged, INumber<T>
```

Comment:
An additional comment here describing the rationale behind using a combination of fast & slow filters would be helpful here